purepyvmware/lib/datastore_utils.py

```diff
@@ -0,0 +1,175 @@
+
+    Returns:
+        pure_devices (list): List of confirmed Pure Storage backed devices.
+    """
+    # Examples of valid device canonical name formats:
+    # naa.624a93703b7b308d98f9425e000113e9
+    # eui.003b7b308d98f94224a9375e00018816
+
+    pure_device = None
+    if device.startswith('naa'):
+        if device[5:11] == self.pure_identifier:
+            pure_device = device
+    elif device.startswith('eui'):
+        if device[20:26] == self.pure_identifier:
+            pure_device = device
+    return pure_device
+
+
+def fa_host_to_esxi_mapping(esxi_host, fa_hosts, flasharray):
+    """Map an ESXi host to a Pure Storage FlashArray host object.
+
+    Args:
+        esxi_host (vim.HostSystem): ESXi host to verify against FA host objects.
+        fa_hosts (list): List of FlashArray host objects and their properties.
+        flasharray (purestorage.FlashArray): Verified or unverified session to
+        a Pure Storage FlashArray.
+
+    Returns:
+        match (dict): FlashArray host object and properties.
+
+    Raises:
+        ValueError: If the ESXi host is unable to be matched to any host object.
+
+    """
+    match = None
+
+    hbas = esxi_host.config.storageDevice.hostBusAdapter
+
+    for hba in hbas:
+        if isinstance(hba, vim.host.FibreChannelHba):
+            # We need to remove the '0x' to properly match against the FA port.
+            hba_identifier = hex(hba.portWorldWideName)[2:]
+        elif isinstance(hba, vim.host.InternetScsiHba):
+            hba_identifier = hba.iScsiName
+        else:
+            # VMware doesn't have the API to get the NVMe NQN yet but this code
+            # will be updated once it is available.
+            continue
+
+        if hba_identifier[0].isdigit():
+            for host in fa_hosts:
+                for wwpn in host['wwn']:
+                    if hba_identifier.lower() == wwpn.lower():
+                        match = host
+                        break
+        elif hba_identifier.startswith('iqn'):
+            for host in fa_hosts:
+                for iqn in host['iqn']:
+                    if hba_identifier.lower() == iqn.lower():
+                        match = host
+                        break
+    if not match:
+        raise ValueError(f'No host object could be found on Pure Storage Flasharray '
+                         f'\'{flasharray.get()["array_name"]}\' for ESXi host \'{esxi_host.name}\'.')
+
+    return match
+
+
+def fa_hgroup_to_vc_cluster_mapping(vc_cluster, flasharray):
+    """Map a vSphere Cluster to a FlashArray host group object.
+
+    Args:
+        vc_cluster (vim.ClusterComputeResource): vSphere cluster to verify
+        against the Pure Storage FlashArray host group object.
+
+    Returns:
+
+
+    Raises:
+        ValueError: If a matched FlashArray host object is not associated with
+        a host group.
+    """
+    fa_name = flasharray.get()["array_name"]
```

Comment:
use .get()

Reply:
@richimus123 -- I am not sure I understand what you are pointing me to here. I am using `.get()`. What am I not understanding?

-jhop

Reply:
Nevermind, I spoke with Fritz about it and he pointed out why I was confused. I forget the `get()` is a custom for the FlashArray Python Toolkit. It has been resolved.

-jhop